File: multi_language/session_raid_stats/session_raid_stats.go
// session_raid_stats.go — Go port of the Session Raid Stats Pine Script indicator
// ================================================================================
// Tracks up to three intraday sessions, detects range-extension raids, buckets
// raid sizes, and computes empirical reach-probabilities for each level.
//
// Build & run:
//   go run session_raid_stats.go

package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const (
	MaxBuckets = 6
	BC         = MaxBuckets + 1 // 7 buckets: 0-5 + overflow
)

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

// SessionConfig holds parameters for one session.
type SessionConfig struct {
	SessionStartH int     // hour (local exchange time)
	SessionStartM int     // minute
	SessionEndH   int
	SessionEndM   int
	MinRaidPts    float64
	CutoffMins    int     // 0 = no cutoff
	BucketStart   float64
	BucketStep    float64
	TZOffsetHrs   float64 // e.g. -5.0 for ET (simplified, no DST)
}

// LevelPts returns the threshold in points for level n (1-indexed).
func (c *SessionConfig) LevelPts(n int) float64 {
	return c.BucketStart + float64(n-1)*c.BucketStep
}

// Levels returns all six level thresholds.
func (c *SessionConfig) Levels() [MaxBuckets]float64 {
	var lvls [MaxBuckets]float64
	for i := 0; i < MaxBuckets; i++ {
		lvls[i] = c.LevelPts(i + 1)
	}
	return lvls
}

// ---------------------------------------------------------------------------
// Bucket helpers
// ---------------------------------------------------------------------------

// BucketIndex maps a raid-size value to a 0-based bucket index.
func BucketIndex(value float64, cfg *SessionConfig) int {
	lvls := cfg.Levels()
	for i := 0; i < MaxBuckets-1; i++ {
		if value < lvls[i+1] {
			return i
		}
	}
	if value < lvls[MaxBuckets-1]+cfg.BucketStep {
		return MaxBuckets - 1
	}
	return MaxBuckets // overflow
}

// CumulativeCounts computes the reverse-cumulative sum.
func CumulativeCounts(counts [BC]int) [BC]int {
	var cum [BC]int
	running := 0
	for i := BC - 1; i >= 0; i-- {
		running += counts[i]
		cum[i] = running
	}
	return cum
}

// UpdateProbCache recalculates ECDF probabilities.
func UpdateProbCache(hiCounts, loCounts [BC]int, dayCount int, probHi, probLo *[BC]float64) {
	if dayCount <= 0 {
		return
	}
	cumH := CumulativeCounts(hiCounts)
	cumL := CumulativeCounts(loCounts)
	for i := 0; i < BC; i++ {
		probHi[i] = float64(cumH[i]) / float64(dayCount) * 100.0
		probLo[i] = float64(cumL[i]) / float64(dayCount) * 100.0
	}
}

// ---------------------------------------------------------------------------
// Session time helpers
// ---------------------------------------------------------------------------

func barLocalTime(timeMS int64, tzOffsetHrs float64) time.Time {
	offsetSec := int64(tzOffsetHrs * 3600)
	return time.Unix(timeMS/1000+offsetSec, 0).UTC()
}

func inSession(timeMS int64, cfg *SessionConfig) bool {
	t     := barLocalTime(timeMS, cfg.TZOffsetHrs)
	h, m  := t.Hour(), t.Minute()
	mins  := h*60 + m
	start := cfg.SessionStartH*60 + cfg.SessionStartM
	end   := cfg.SessionEndH*60 + cfg.SessionEndM
	return mins >= start && mins < end
}

func withinCutoff(timeMS, rangeEndMS int64, cutoffMins int) bool {
	if cutoffMins == 0 {
		return true
	}
	return (timeMS - rangeEndMS) <= int64(cutoffMins)*60_000
}

// ---------------------------------------------------------------------------
// Raid engine
// ---------------------------------------------------------------------------

// RaidState holds per-session mutable tracking state.
type RaidState struct {
	Hi, Lo            float64
	EndMS             int64
	Active            bool
	HiMax             float64
	HiTouch, HiConf   bool
	HiPts             float64
	LoMin             float64
	LoTouch, LoConf   bool
	LoPts             float64
	CutExp            bool
	hiSet, loSet      bool
}

// RaidStats holds the computed statistics for a session.
type RaidStats struct {
	ProbHi   [BC]float64
	ProbLo   [BC]float64
	DayCount int
}

// RaidEngine processes chronological OHLC bars for one session.
type RaidEngine struct {
	cfg        *SessionConfig
	maxDays    int
	state      RaidState
	hiCounts   [BC]int
	loCounts   [BC]int
	probHi     [BC]float64
	probLo     [BC]float64
	dayCount   int
	prevInSess bool
}

// NewRaidEngine creates a new engine for the given configuration.
func NewRaidEngine(cfg *SessionConfig, maxDays int) *RaidEngine {
	return &RaidEngine{cfg: cfg, maxDays: maxDays}
}

// OnBar feeds one OHLC bar to the engine.
func (e *RaidEngine) OnBar(timeMS int64, open, high, low, close float64) {
	inS   := inSession(timeMS, e.cfg)
	state := &e.state

	// Session open — reset state
	if inS && !e.prevInSess {
		*state = RaidState{Hi: high, Lo: low, hiSet: true, loSet: true}
	}

	// Expand range during session
	if inS && state.hiSet {
		if high > state.Hi { state.Hi = high }
		if low  < state.Lo { state.Lo = low  }
	}

	// Session close — activate raid detection
	if !inS && e.prevInSess && state.hiSet {
		state.EndMS  = timeMS
		state.Active = true
		if e.dayCount < e.maxDays {
			e.dayCount++
		}
	}

	// Raid detection
	if state.Active {
		if withinCutoff(timeMS, state.EndMS, e.cfg.CutoffMins) {
			// High-side
			if high > state.Hi {
				ext := high - state.Hi
				if ext >= e.cfg.MinRaidPts {
					if !state.HiTouch || high > state.HiMax {
						state.HiMax   = high
						state.HiPts   = ext
						state.HiTouch = true
					}
				}
			}
			// Low-side
			if low < state.Lo {
				ext := state.Lo - low
				if ext >= e.cfg.MinRaidPts {
					if !state.LoTouch || low < state.LoMin {
						state.LoMin   = low
						state.LoPts   = ext
						state.LoTouch = true
					}
				}
			}
		} else if !state.CutExp {
			if state.HiTouch {
				idx := BucketIndex(state.HiPts, e.cfg)
				e.hiCounts[idx]++
				state.HiConf = true
			}
			if state.LoTouch {
				idx := BucketIndex(state.LoPts, e.cfg)
				e.loCounts[idx]++
				state.LoConf = true
			}
			UpdateProbCache(e.hiCounts, e.loCounts, e.dayCount, &e.probHi, &e.probLo)
			state.CutExp = true
		}
	}

	e.prevInSess = inS
}

// GetStats returns a snapshot of the current statistics.
func (e *RaidEngine) GetStats() RaidStats {
	return RaidStats{ProbHi: e.probHi, ProbLo: e.probLo, DayCount: e.dayCount}
}

// ---------------------------------------------------------------------------
// Three-session facade
// ---------------------------------------------------------------------------

// SessionRaidStats wraps three RaidEngines.
type SessionRaidStats struct {
	E1, E2, E3 *RaidEngine
}

// NewSessionRaidStats constructs the three-engine facade.
func NewSessionRaidStats(c1, c2, c3 *SessionConfig, maxDays int) *SessionRaidStats {
	return &SessionRaidStats{
		E1: NewRaidEngine(c1, maxDays),
		E2: NewRaidEngine(c2, maxDays),
		E3: NewRaidEngine(c3, maxDays),
	}
}

// OnBar feeds a bar to all three engines.
func (s *SessionRaidStats) OnBar(timeMS int64, o, h, l, c float64) {
	s.E1.OnBar(timeMS, o, h, l, c)
	s.E2.OnBar(timeMS, o, h, l, c)
	s.E3.OnBar(timeMS, o, h, l, c)
}

// ---------------------------------------------------------------------------
// main — quick self-test
// ---------------------------------------------------------------------------

func main() {
	cfg := &SessionConfig{
		SessionStartH: 2, SessionStartM: 0,
		SessionEndH:   2, SessionEndM:   15,
		MinRaidPts:    5.0, CutoffMins: 120,
		BucketStart:   20.0, BucketStep: 10.0,
		TZOffsetHrs:   -5.0,
	}
	engine := NewRaidEngine(cfg, 100)

	rng := rand.New(rand.NewSource(42))
	// 2024-01-02 07:00 UTC = 02:00 ET
	baseMS := int64(1704182400000)
	msMin  := int64(60000)
	price  := 4500.0

	for day := 0; day < 5; day++ {
		dayOff := int64(day) * 24 * 60 * msMin
		for minute := 0; minute < 8*60; minute++ {
			t := baseMS + dayOff + int64(minute)*msMin
			o := price
			h := o + rng.Float64()*4
			l := o - rng.Float64()*4
			c := l + rng.Float64()*(h-l)
			price = c
			engine.OnBar(t, o, h, l, c)
		}
	}

	stats := engine.GetStats()
	fmt.Printf("Day count : %d\n", stats.DayCount)
	fmt.Print("prob_hi   :")
	for _, p := range stats.ProbHi {
		fmt.Printf(" %.1f%%", p)
	}
	fmt.Print("\nprob_lo   :")
	for _, p := range stats.ProbLo {
		fmt.Printf(" %.1f%%", math.Round(p*10)/10)
	}
	fmt.Println()
}

File: multi_language/session_raid_stats/SessionRaidStats.java
/**
 * SessionRaidStats.java — Java 11+ port of the Session Raid Stats Pine Script indicator
 * =======================================================================================
 * Tracks up to three intraday sessions, detects range-extension raids, buckets
 * raid sizes, and computes empirical reach-probabilities for each level.
 *
 * Compile & run:
 *   javac SessionRaidStats.java && java SessionRaidStats
 */

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Arrays;
import java.util.Random;

public class SessionRaidStats {

    // -----------------------------------------------------------------------
    // Constants
    // -----------------------------------------------------------------------

    static final int MAX_BUCKETS = 6;
    static final int BC          = MAX_BUCKETS + 1; // 7 buckets: 0-5 + overflow

    // -----------------------------------------------------------------------
    // Configuration
    // -----------------------------------------------------------------------

    static class SessionConfig {
        final int     sessionStartH;
        final int     sessionStartM;
        final int     sessionEndH;
        final int     sessionEndM;
        final double  minRaidPts;
        final int     cutoffMins;   // 0 = no cutoff
        final double  bucketStart;
        final double  bucketStep;
        final double  tzOffsetHrs;  // simplified, no DST

        SessionConfig(int startH, int startM, int endH, int endM,
                      double minRaidPts, int cutoffMins,
                      double bucketStart, double bucketStep, double tzOffsetHrs) {
            this.sessionStartH = startH;  this.sessionStartM = startM;
            this.sessionEndH   = endH;    this.sessionEndM   = endM;
            this.minRaidPts    = minRaidPts;
            this.cutoffMins    = cutoffMins;
            this.bucketStart   = bucketStart;
            this.bucketStep    = bucketStep;
            this.tzOffsetHrs   = tzOffsetHrs;
        }

        double levelPts(int n) {
            return bucketStart + (n - 1) * bucketStep;
        }

        double[] levels() {
            double[] lvls = new double[MAX_BUCKETS];
            for (int i = 0; i < MAX_BUCKETS; i++) lvls[i] = levelPts(i + 1);
            return lvls;
        }
    }

    // -----------------------------------------------------------------------
    // Bucket helpers
    // -----------------------------------------------------------------------

    static int bucketIndex(double value, SessionConfig cfg) {
        double[] lvls = cfg.levels();
        for (int i = 0; i < MAX_BUCKETS - 1; i++) {
            if (value < lvls[i + 1]) return i;
        }
        if (value < lvls[MAX_BUCKETS - 1] + cfg.bucketStep) return MAX_BUCKETS - 1;
        return MAX_BUCKETS; // overflow
    }

    static int[] cumulativeCounts(int[] counts) {
        int[] cum = new int[counts.length];
        int running = 0;
        for (int i = counts.length - 1; i >= 0; i--) {
            running += counts[i];
            cum[i] = running;
        }
        return cum;
    }

    static void updateProbCache(int[] hiCounts, int[] loCounts, int dayCount,
                                 double[] probHi, double[] probLo) {
        if (dayCount <= 0) return;
        int[] cumH = cumulativeCounts(hiCounts);
        int[] cumL = cumulativeCounts(loCounts);
        for (int i = 0; i < BC; i++) {
            probHi[i] = (double) cumH[i] / dayCount * 100.0;
            probLo[i] = (double) cumL[i] / dayCount * 100.0;
        }
    }

    // -----------------------------------------------------------------------
    // Session time helpers
    // -----------------------------------------------------------------------

    static int barLocalMins(long timeMs, double tzOffsetHrs) {
        long offsetMs = (long)(tzOffsetHrs * 3_600_000L);
        Instant inst  = Instant.ofEpochMilli(timeMs + offsetMs);
        ZonedDateTime zdt = inst.atZone(ZoneOffset.UTC);
        return zdt.getHour() * 60 + zdt.getMinute();
    }

    static boolean inSession(long timeMs, SessionConfig cfg) {
        int mins  = barLocalMins(timeMs, cfg.tzOffsetHrs);
        int start = cfg.sessionStartH * 60 + cfg.sessionStartM;
        int end   = cfg.sessionEndH   * 60 + cfg.sessionEndM;
        return mins >= start && mins < end;
    }

    static boolean withinCutoff(long timeMs, long rangeEndMs, int cutoffMins) {
        if (cutoffMins == 0) return true;
        return (timeMs - rangeEndMs) <= (long) cutoffMins * 60_000L;
    }

    // -----------------------------------------------------------------------
    // Raid state (mutable per-session)
    // -----------------------------------------------------------------------

    static class RaidState {
        double hi = Double.NaN, lo = Double.NaN;
        long   endMs   = Long.MIN_VALUE;
        boolean active = false;
        double hiMax   = Double.NaN;
        boolean hiTouch = false, hiConf = false;
        double  hiPts   = Double.NaN;
        double  loMin   = Double.NaN;
        boolean loTouch = false, loConf = false;
        double  loPts   = Double.NaN;
        boolean cutExp  = false;

        void reset(double firstHigh, double firstLow) {
            hi = firstHigh; lo = firstLow;
            endMs = Long.MIN_VALUE; active = false;
            hiMax = Double.NaN; hiTouch = false; hiConf = false; hiPts = Double.NaN;
            loMin = Double.NaN; loTouch = false; loConf = false; loPts = Double.NaN;
            cutExp = false;
        }
    }

    // -----------------------------------------------------------------------
    // Raid stats snapshot
    // -----------------------------------------------------------------------

    static class RaidStatsSnapshot {
        final double[] probHi;
        final double[] probLo;
        final int      dayCount;

        RaidStatsSnapshot(double[] probHi, double[] probLo, int dayCount) {
            this.probHi   = Arrays.copyOf(probHi, probHi.length);
            this.probLo   = Arrays.copyOf(probLo, probLo.length);
            this.dayCount = dayCount;
        }
    }

    // -----------------------------------------------------------------------
    // Engine
    // -----------------------------------------------------------------------

    static class RaidEngine {
        private final SessionConfig cfg;
        private final int           maxDays;
        private final RaidState     state    = new RaidState();
        private final int[]         hiCounts = new int[BC];
        private final int[]         loCounts = new int[BC];
        private final double[]      probHi   = new double[BC];
        private final double[]      probLo   = new double[BC];
        private int     dayCount   = 0;
        private boolean prevInSess = false;

        RaidEngine(SessionConfig cfg, int maxDays) {
            this.cfg     = cfg;
            this.maxDays = maxDays;
        }

        void onBar(long timeMs, double open, double high, double low, double close) {
            boolean inS = inSession(timeMs, cfg);

            // Session open
            if (inS && !prevInSess) {
                state.reset(high, low);
            }

            // Expand range
            if (inS && !Double.isNaN(state.hi)) {
                if (high > state.hi) state.hi = high;
                if (low  < state.lo) state.lo = low;
            }

            // Session close
            if (!inS && prevInSess && !Double.isNaN(state.hi)) {
                state.endMs  = timeMs;
                state.active = true;
                if (dayCount < maxDays) dayCount++;
            }

            // Raid detection
            if (state.active) {
                if (withinCutoff(timeMs, state.endMs, cfg.cutoffMins)) {
                    if (high > state.hi) {
                        double ext = high - state.hi;
                        if (ext >= cfg.minRaidPts && (Double.isNaN(state.hiMax) || high > state.hiMax)) {
                            state.hiMax   = high;
                            state.hiPts   = ext;
                            state.hiTouch = true;
                        }
                    }
                    if (low < state.lo) {
                        double ext = state.lo - low;
                        if (ext >= cfg.minRaidPts && (Double.isNaN(state.loMin) || low < state.loMin)) {
                            state.loMin   = low;
                            state.loPts   = ext;
                            state.loTouch = true;
                        }
                    }
                } else if (!state.cutExp) {
                    if (state.hiTouch) {
                        hiCounts[bucketIndex(state.hiPts, cfg)]++;
                        state.hiConf = true;
                    }
                    if (state.loTouch) {
                        loCounts[bucketIndex(state.loPts, cfg)]++;
                        state.loConf = true;
                    }
                    updateProbCache(hiCounts, loCounts, dayCount, probHi, probLo);
                    state.cutExp = true;
                }
            }

            prevInSess = inS;
        }

        RaidStatsSnapshot getStats() {
            return new RaidStatsSnapshot(probHi, probLo, dayCount);
        }
    }

    // -----------------------------------------------------------------------
    // Three-session facade
    // -----------------------------------------------------------------------

    private final RaidEngine e1, e2, e3;

    public SessionRaidStats(SessionConfig c1, SessionConfig c2, SessionConfig c3, int maxDays) {
        e1 = new RaidEngine(c1, maxDays);
        e2 = new RaidEngine(c2, maxDays);
        e3 = new RaidEngine(c3, maxDays);
    }

    public void onBar(long timeMs, double open, double high, double low, double close) {
        e1.onBar(timeMs, open, high, low, close);
        e2.onBar(timeMs, open, high, low, close);
        e3.onBar(timeMs, open, high, low, close);
    }

    public RaidStatsSnapshot[] stats() {
        return new RaidStatsSnapshot[]{ e1.getStats(), e2.getStats(), e3.getStats() };
    }

    // -----------------------------------------------------------------------
    // main — quick self-test
    // -----------------------------------------------------------------------

    public static void main(String[] args) {
        SessionConfig cfg = new SessionConfig(2, 0, 2, 15,
            5.0, 120, 20.0, 10.0, -5.0);
        RaidEngine engine = new RaidEngine(cfg, 100);

        Random rng     = new Random(42);
        long   baseMs  = 1704182400_000L; // 2024-01-02 07:00 UTC = 02:00 ET
        long   msMin   = 60_000L;
        double price   = 4500.0;

        for (int day = 0; day < 5; day++) {
            long dayOff = (long) day * 24 * 60 * msMin;
            for (int m = 0; m < 8 * 60; m++) {
                long   t = baseMs + dayOff + (long) m * msMin;
                double o = price;
                double h = o + rng.nextDouble() * 4;
                double l = o - rng.nextDouble() * 4;
                double c = l + rng.nextDouble() * (h - l);
                price = c;
                engine.onBar(t, o, h, l, c);
            }
        }

        RaidStatsSnapshot stats = engine.getStats();
        System.out.println("Day count : " + stats.dayCount);
        System.out.print("prob_hi   :");
        for (double p : stats.probHi) System.out.printf(" %.1f%%", p);
        System.out.print("\nprob_lo   :");
        for (double p : stats.probLo) System.out.printf(" %.1f%%", p);
        System.out.println();
    }
}
